### PR TITLE
Make where we work with BOSH Release tarballs explicit

### PR DESCRIPTION
**Pull request by crhntr**
## TL;DR

- `component.Spec` and `cargo.ComponentSpec` -> `cargo.BOSHReleaseTarballSpecification`
- `component.Lock` and `cargo.ComponentLock` -> `cargo.BOSHReleaseTarballLock`
- `cargo.ReleaseSourceTypeBOSHIO` -> `cargo.BOSHReleaseTarballSourceTypeBOSHIO`
- `cargo.ReleaseSourceTypeS3` -> `cargo.BOSHReleaseTarballSourceTypeS3`
- `cargo.ReleaseSourceTypeGithub` -> `cargo.BOSHReleaseTarballSourceTypeGithub`
- `cargo.ReleaseSourceTypeArtifactory` -> `cargo.BOSHReleaseTarballSourceTypeArtifactory`
- *a few other function and method names were changed in a similar way to align with the above rename operations.*

## Discussion

This refactor is intended to make it easier to think about and code around the BOSH releases tarballs we package.

- In 96032c82cec94316a3784b5960ded477afd2a761, I merge `component.Lock` and `cargo.ComponentLock`. They have [literally been the same thing for a while](https://github.com/pivotal-cf/kiln/blob/57a648139d891c36ea43a113d22545bb3efda0f3/internal/component/component.go#L9-L22); this commit removes the indirection.
  - `component.Local` and `component.Exported` no longer embed the `cargo.ComponentLock` but have it as a non-embedded field. This makes refactors easier and makes it clear that those types should not inherit methods from the Lock. For the most part, this change is reversible in a non-breaking way.
- In a67fdfd0fd92cf5db8738b42d644f0efea7092a2, I rename "Component{Lock,Spec} to BOSHRelease{Lock,Spec}. We may still want to refer to BOSH Releases as tile components in speech and docs but the code reads better when it is less ambiguous about what is is operating on.
- In f0f1656d3c38e4ad1f99bf15f8b5245b63c38bf1, I add the Tarball part to the BOSHRelease names.
- In 797771017b28b27e6f8775ef77180e6031a76cd9, I renamed other public identifiers and methods to stay consistent with the Tarball Lock and Spec types.

### CODE REVIEW

**Comment by crhntr** [comment]
[Prior art](https://github.com/pivotal-cf/kiln/pull/328/commits/4fc5bf46a2e927d3178132d55e895d13f1cd5676)

**Review by dtimm — APPROVED** [review]
Excellent refactor. LGTM.
